src/pp_config_v2.erl

```diff
@@ -0,0 +1,267 @@
+    #{
+        <<"name">> => Name,
+        <<"net_id">> => NetID,
+        %% old config won't have devaddrs key
+        %% and might not have joins key
+        <<"configs">> => [
+            maps:merge(
+                #{
+                    <<"devaddrs">> => [],
+                    <<"joins">> => []
+                },
+                Inner
+            )
+        ]
+    }.
+
+-spec eui_from_configs(binary(), integer(), list(map())) -> list(#eui{}).
+eui_from_configs(Name, NetID, Configs) ->
+    lists:flatten(
+        lists:map(
+            fun(Entry) ->
+                #{<<"joins">> := EUIs} = Entry,
+                BuyingActive = get_buying_active(Entry),
+                MultiBuy = get_multi_buy(Entry),
+                Protocol = get_protocol(Entry),
+                lists:map(
+                    fun(#{<<"dev_eui">> := DevBin, <<"app_eui">> := AppBin}) ->
+                        #eui{
+                            name = Name,
+                            net_id = NetID,
+                            app_eui = hex_to_num(AppBin),
+                            dev_eui = hex_to_num(DevBin),
+                            multi_buy = MultiBuy,
+                            protocol = Protocol,
+                            buying_active = BuyingActive
+                        }
+                    end,
+                    EUIs
+                )
+            end,
+            Configs
+        )
+    ).
+
+-spec get_buying_active(map()) -> boolean().
+get_buying_active(#{<<"active">> := Active}) -> Active;
+get_buying_active(_) -> ?DEFAULT_ACTIVE.
+
+-spec get_multi_buy(map()) -> unlimited | non_neg_integer().
+get_multi_buy(#{<<"multi_buy">> := null}) -> ?DEFAULT_MULTI_BUY;
+get_multi_buy(#{<<"multi_buy">> := <<"unlimited">>}) -> unlimited;
+get_multi_buy(#{<<"multi_buy">> := MB}) -> MB;
+get_multi_buy(_) -> ?DEFAULT_MULTI_BUY.
+
+-spec devaddr_from_configs(binary(), integer(), list(map())) -> list(#devaddr{}).
+devaddr_from_configs(Name, NetID, Configs) ->
+    lists:flatten(
+        lists:map(
+            fun(Entry) ->
+                #{<<"devaddrs">> := DevAddrs} = Entry,
+                BuyingActive = maps:get(<<"active">>, Entry, ?DEFAULT_ACTIVE),
+                MultiBuy = get_multi_buy(Entry),
+                Protocol = get_protocol(Entry),
+                lists:map(
+                    fun(#{<<"lower">> := Lower, <<"upper">> := Upper}) ->
+                        #devaddr{
+                            name = Name,
+                            net_id = NetID,
+                            protocol = Protocol,
+                            multi_buy = MultiBuy,
+                            buying_active = BuyingActive,
+                            addr = {range, hex_to_num(Lower), hex_to_num(Upper)}
+                        }
+                    end,
+                    case DevAddrs of
+                        [] -> [#{<<"lower">> => <<"0x00000000">>, <<"upper">> => <<"0xFFFFFFFF">>}];
```

Comment:
The 0x Binary values aren't Erlang hex.  If that's intended, please add a comment explaining.

(I see that you have `hex_to_num()` below where you parse the `<<"0x">>` prefix, but on first read here, it looks like a possible typo.)

Reply:
I'll add a good note when I update the docs.
All the "hex" values in packet purchaser right now are roaming1.1 spec "hex" values.